app/worker/worker_group.go

```diff
@@ -0,0 +1,242 @@
+
+	"github.com/mendersoftware/go-lib-micro/log"
+
+	"github.com/mendersoftware/workflows/client/nats"
+	"github.com/mendersoftware/workflows/model"
+	"github.com/mendersoftware/workflows/store"
+)
+
+type workerGroup struct {
+	mu        sync.Mutex    // Mutex to protect the shared firstDone channel
+	firstDone chan struct{} // First worker finished
+	done      chan struct{} // All workers finished
+	termID    int32         // ID of the first worker to finish
+
+	sidecarDone chan struct{}
+	workerCount int32
+
+	input        <-chan *natsio.Msg
+	notifyPeriod time.Duration
+
+	client nats.Client
+	store  store.DataStore
+}
+
+func NewWorkGroup(
+	input <-chan *natsio.Msg,
+	notifyPeriod time.Duration,
+	nc nats.Client,
+	ds store.DataStore,
+) *workerGroup {
+	return &workerGroup{
+		sidecarDone: make(chan struct{}),
+		done:        make(chan struct{}),
+		firstDone:   make(chan struct{}),
+
+		input:        input,
+		notifyPeriod: notifyPeriod,
+		client:       nc,
+		store:        ds,
+	}
+}
+
+// Done returns a channel (barrier) that is closed when the last worker
+// has exited.
+func (w *workerGroup) Done() <-chan struct{} {
+	return w.done
+}
+
+// FirstDone returns a channel (barrier) that is closed when the first
+// worker has exited.
+func (w *workerGroup) FirstDone() <-chan struct{} {
+	return w.firstDone
+}
+
+// TermID is the ID of the first worker that quit.
+func (w *workerGroup) TermID() int32 {
+	return w.termID
+}
+
+func (w *workerGroup) RunWorker(ctx context.Context) {
+	id := atomic.AddInt32(&w.workerCount, 1)
+	l := log.FromContext(ctx)
+	l = l.F(log.Ctx{"worker_id": id})
+	ctx = log.WithContext(ctx, l)
+
+	sidecarChan := make(chan *natsio.Msg, 1)
+	sidecarDone := make(chan struct{})
+	defer func() {
+		l.Info("worker shutting down")
+		close(sidecarChan)
+
+		w.mu.Lock()
+		remaining := atomic.AddInt32(&w.workerCount, -1)
+		// Is this the last worker to quit?
+		if remaining <= 0 {
+			select {
+			case <-w.done:
+
+			default:
+				close(w.done)
+			}
+		} else {
+			// Is this the first worker to quit?
+			select {
+			case <-w.firstDone:
+
+			default:
+				w.termID = id
+				close(w.firstDone)
+			}
+		}
+		w.mu.Unlock()
+	}()
+	l.Info("worker starting up")
+	// workerSidecar is responsible for notifying the broker about slow workflows
+	go w.workerSidecar(ctx, sidecarChan, sidecarDone)
+	w.workerMain(ctx, sidecarChan, sidecarDone)
+}
+
+func (w *workerGroup) workerMain(
+	ctx context.Context,
+	sidecarChan chan *natsio.Msg,
+	sidecarDone chan struct{},
+) {
+	l := log.FromContext(ctx)
+	ctxDone := ctx.Done()
+	sidecarTimer := (*reusableTimer)(time.NewTimer(0))
+	for {
+		var (
+			msg    *natsio.Msg
+			isOpen bool
+		)
+		select {
+		case msg, isOpen = <-w.input:
+			if !isOpen {
+				return
+			}
+
+		case <-sidecarDone:
+			return
+		case <-ctxDone:
+			return
+		}
+
+		// Notify the sidecar routine about the new message
+		select {
+		case sidecarChan <- msg:
+
+		case <-sidecarTimer.After(w.notifyPeriod / 8):
```

Comment:
`8` here is a magic number? could we define it?

Reply:
:8ball: